### ISSUE TITLE
how to wait for url change after .useRole and then trigger t.navigateTo

### ISSUE THREAD

**Issue by Hotell** [DevExpress/testcafe#1836]
### Are you requesting a feature or reporting a bug?
This is rather a question which if solved should be added to recipes docs.

### What is the current behavior?

I have a SPA, Login form, using testcafe Roles api.

scenario is something like following:

```ts
import {USER_DASHBOARD_URL} from './routes'

const admin = Role(
  LOGIN_URL,
  async t => {
    await t
      .typeText(loginPage.email, 'foooo')
      .typeText(loginPage.password, 'baaar')
      .click(loginPage.submit)
  }
)

fixture('#admin | tests for user dashboard')
  .beforeEach( async t => { 
    await t
       // authenticate before each test so we don't have to do the same wihtin every test
      .useRole(admin)
      // navigate to tested url
      .navigateTo(USER_DASHBOARD_URL)
  })

test('there should be 3 users in table', async t => { 
  await t.expect(table.tableFilterItems.filterCountrySelect.exists).eql(3)
})
```

Step by step process:
```
-> starting URL: #/login
-> useRole applied
  -> click submits form 
-> XHR for authentitacion is executed
  -> on XHR success user is redirected to `#/`
-> navigateTo( testedURL ) - this should be executed after XHR is done and page was redirected to `#/`
```
**problem:**

 time to time, test will stay hanging on login page, because authentication wasn't finished yet, but Testcafe executed `.navigateTo()` already, which won't success cause user is not yet logged in, so app logic will redirect that navigation action back to login. Of course if this occurs once in a test suite batch, all test will fail ( because testcafe remembers "auth role" state )

Only checkmark ( that came up my mind ) is that, when user is successfuly logged in, he is redirected from `/login` to `/`

I really got stuck with this and don't know how to proceed further, only thing that comes up my mind is somehow tell testcafe that it should execute that `.navigateTo` only after `location.pathname` was changed from `/login` -> `/`

Any suggestions are more than welcome :)

### What is the expected behavior?
Provide a solution for described problem

### Specify your

* operating system: OSX 10.12.6
* testcafe version:  0.17.2
* node.js version: v8.4.0

**Comment by AlexanderMoskovkin** [comment]
Hi @Hotell,

TestCafe should wait automatically for any xhr requests and page reloads after each action. It seems something goes wrong in your case. If you can provide us with your page url or a simple example we can try to find the reason of the issue. 
Anyway you can resolve the issue by using assertions:

```js
const getLocation  = ClientFunction(() => document.location.href.toString());

await t
    .useRole(admin)
    .expect(getLocation()).notContains('login')
    .navigateTo(USER_DASHBOARD_URL);
```

Please let me know does this solution work?

**Comment by Hotell** [comment]
hmm that assertion doesn't work 100%.

What is weird though, is sometimes the login passes and then all tests run ok, but sometimes it keeps hanging and all following tests will fail. 

The app is in a private corporate environment so I'm not able to share :(

**Comment by AlexanderMoskovkin** [comment]
> hmm that assertion doesn't work 100%.

What do you mean? The assertion fails? Or it's passed even the login is not completed?

Meanwhile, just for check try to run tests with a lower speed (for exapmle: 0.3-0.5) (see [--speed](https://devexpress.github.io/testcafe/documentation/using-testcafe/command-line-interface.html#--speed-factor)) and please let me know are there any changes?

**Comment by Hotell** [comment]
also the app is not using HTML5_MODE for routing -> just hashes so `localhost:8080/#/login` -> `localhost:8080/#/` etc

**Comment by Hotell** [comment]
with `--speed 0.5` tests are always passing 😕

**Comment by Hotell** [comment]
also forgot to mention that login process, executes 2 XHR
-> click
-> POST login
-> GET user
-> go to `/#/` ( Dashboard )

**Comment by AlexanderMoskovkin** [comment]
Are they sent sequentially? The page gets a response for the first request and then sends the second?

**Comment by Hotell** [comment]
yes sequentially. yup

when I set speed to 1 some test passes and some error out on that "hack" assertion

![image](https://user-images.githubusercontent.com/1223799/31085057-6f0b5f54-a796-11e7-9129-c1b9cbda056f.png)

**Comment by AlexanderMoskovkin** [comment]
Thanks for the provided info. Could you please do one more check. 

First, add `await t.wait(2000)` before `.useRole(admin)`. Run the tests with default speed (1) and see are the tests ok?

```js
await t
      .wait(2000)
      .useRole(admin)
      .navigateTo(USER_DASHBOARD_URL)
```

Then place the `wait` command after `userRole(admin)`. Again run with the speed `1`. How it's now?

```js
await t
      .useRole(admin)
      .wait(2000)
      .navigateTo(USER_DASHBOARD_URL)
```

It'll help me to find a possible reason of the issue because I have two different assumptions now.

**Comment by Hotell** [comment]
For further reference, here is the auth function that I'm using
```ts
const authenticate = async ({ username, password }: UserCredentials, t: TestController) => {
  return await t
    .wait(2000)
    .typeText(loginPage.email, username)
    .typeText(loginPage.password, password)
    .click(loginPage.submit)
    // hack https://github.com/DevExpress/testcafe/issues/1836#issuecomment-333476931
    .expect(getLocation())
    .notContains(ROOT_PATH)
    .navigateTo(DASHBOARD_URL)
}

export const admin = Role(LOGIN_URL, async t => {
  await authenticate({ username: '----', password: '---' }, t)
})
```

- wait before
**EVERYTHING IS PASSING on 1st run**
![image](https://user-images.githubusercontent.com/1223799/31089117-2ec73c62-a7a3-11e7-93f2-0a5d851b0edf.png)

**REPEATED RUN FAILS**
![image](https://user-images.githubusercontent.com/1223799/31089557-e501fa7a-a7a4-11e7-8f5f-13c0247282f4.png)


- wait after
**FAILED**
![image](https://user-images.githubusercontent.com/1223799/31089286-c954bfd4-a7a3-11e7-89ef-b4233970607f.png)

![image](https://user-images.githubusercontent.com/1223799/31089261-b927626a-a7a3-11e7-9f60-52ea0a502ef2.png)

**Comment by AlexanderMoskovkin** [comment]
Since we have no access to your page I can assume the following reasons:
- The page is not completely loaded/initialized before typing (the `wait` before typing should help in this case);
- Xhr-requests after the click on the submit button are sent with some delay and TestCafe doesn't wait them before `navigateTo` (the `wait` after click should help there);
- Some page scripts should be executed between typing/click actions on the login form. TestCafe performs actions rapidly so the scripts are not executed properly (lower test speed should help there).

As you say, `wait` actions didn't help. You can try to use the [setTestSpeed](https://devexpress.github.io/testcafe/documentation/test-api/test-code-structure.html#setting-test-speed) setting inside the Role initialization function:

```js
const authenticate = async ({ username, password }: UserCredentials, t: TestController) => {
  return await t
    .setTestSpeed(0.5)
    .typeText(loginPage.email, username)
    .typeText(loginPage.password, password)
    .click(loginPage.submit)
    // hack https://github.com/DevExpress/testcafe/issues/1836#issuecomment-333476931
    .expect(getLocation())
    .notContains(ROOT_PATH)
    .navigateTo(DASHBOARD_URL)
}
```

It will not affect the speed of the whole test run speed a lot because the speed will be decreased only for actions inside the `Role` initialization function which is executed once for the test run.

**Comment by Hotell** [comment]
@AlexanderMoskovkin thanks so much, that works!

**Comment by Hotell** [comment]
I have though other "bug" issue, with resolving `import` , I'll create another ticket for that. I guess this can be closed for now, WDYT ?

**Comment by AlexanderMoskovkin** [comment]
Great to hear that it works! Yep, we can close it.
